ROX-10304: Remove NoValidRole  error sentinel

mtodor opened the pull request:
## Description

Removed `GenericNoValidRole` error. We have added new Error sentinel in `pkg/auth/role`.

**Note:** This PR is based on `mtodor/ROX-10303-replace-new-error-wrappers` branch. After PR #1389 is merged, we will change the base branch.

## Checklist
- [x] Investigated and inspected CI test results
- ~~[ ] Unit test and regression tests added~~ - nothing is added
- ~~[ ] Evaluated and added CHANGELOG entry if required~~ - not relevant
- ~~[ ] Determined and documented upgrade steps~~ - nothing relevant for upgrade steps
- ~~[ ] Documented user facing changes (create PR based on [stackrox/openshift-docs](https://github.com/stackrox/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~~ - no user-facing changes

## Testing Performed

**1. Compare in UI - No change** - *no change*
![Screenshot 2022-04-25 at 12 01 28](https://user-images.githubusercontent.com/3965286/165098428-6ccc855b-32bf-4ee8-84aa-325d832788a8.png)

**2. REST request with token without Role** - *no change*
```
{"error":"access for this user is not authorized: no valid role, please contact your system administrator","code":16,"message":"access for this user is not authorized: no valid role, please contact your system administrator","details":[]}
```

**3. Request with not Admin role:** - *no change*
```
❯ ./bin/darwin/roxctl --insecure-skip-tls-verify --insecure --endpoint https://localhost:8000 --token-file ./token.txt  central backup
ERROR:  Invalid credentials. Please add/fix your credentials

❯ ./bin/darwin/roxctl --insecure-skip-tls-verify --insecure --endpoint https://localhost:8000 --token-file ./token.txt  central debug log
ERROR:	could not get log level from central: rpc error: code = Unauthenticated desc = access for this user is not authorized: no valid role, please contact your system administrator
```

Additionally, default checks are also done:
- `make image` works
- `make style` works so that the linting of all go files is correct

janisz commented on pkg/grpc/authn/interceptor.go:
```diff
@@ -33,7 +33,7 @@ func (u contextUpdater) updateContext(ctx context.Context) (context.Context, err
 		// Only service identities can have no roles assigned.
 		if len(id.Roles()) == 0 && id.Service() == nil {
-			return context.WithValue(ctx, identityErrorContextKey{}, errox.GenericNoValidRole()), nil
+			return context.WithValue(ctx, identityErrorContextKey{}, errox.NotAuthorized.CausedBy("no valid role")), nil
```
```suggestion
			return context.WithValue(ctx, identityErrorContextKey{}, errox.NotAuthorized.CausedBy("only service identities can have no roles assigned")), nil
```

janisz commented on pkg/grpc/authz/user/role.go:
```diff
@@ -30,7 +30,7 @@ func (p *roleChecker) Authorized(ctx context.Context, _ string) error {
 func (p *roleChecker) checkRole(roleNames []string) error {
 	if len(roleNames) == 0 {
-		return errox.NoValidRole
+		return errox.NotAuthorized
```
how about adding more context here e.g. 
```suggestion
		return errox.NotAuthorized.CausedBy("no roles assigned to the user"))
```

janisz commented on pkg/auth/authproviders/registry_httphandler.go:
```diff
@@ -279,9 +279,10 @@ func (r *registryImpl) providersHTTPHandler(w http.ResponseWriter, req *http.Req
 	userRoles := userInfo.GetRoles()
 	if len(userRoles) == 0 {
-		err := errox.GenericNoValidRole()
+		err := errox.NotAuthorized.CausedBy("no valid role for the user")
```
How about adding explicit information about missing roles. `no valid role` sounds like there are some roles but not valid for this situation
```suggestion
		err := errox.NotAuthorized.CausedBy(fmt.Sprintf("%s has no roles", userInfo.GetFriendlyName()))
```

rukletsov reviewed (changes requested):
I believe `401` is the right choice of error here, see https://issues.redhat.com/browse/ROX-8092 including comments. cc @rhybrillou 

I'd suggest to introduce a new sentinel error in `pkg/grpc/auth`, maybe in `none.go` or in a new file, that will replace both `NoValidRole` and `GenericNoValidRole()`, e.g. 
```
ErrNoValidRole := errox.NoCredentials.New"access for this user is not authorized: no valid role," +
		" please contact your system administrator"
```

mtodor commented:
@rukletsov I have applied your suggestion. It looks cleaner. I have added it into `pkg/auth/role` because we are returning this error also in `pkg/auth/authproviders`.

@janisz Since we have changed implementation, your comments are obsolete, and I'll not address them. Please check the new implementation.

rukletsov commented on pkg/auth/role/error.go:
```diff
@@ -0,0 +1,11 @@
+package role
```
Maybe put this file directly into `pkg/auth/`? This way we avoid having a `auth/role` package with just an error definition and the call sites look sleeker: `auth.ErrNoValidRole`.

mtodor replied:
✅

md2119 reviewed (approved):
The changes are fine, but is there an easy way for developers to look for sentinal values, especially if they are scattered in different pkgs?

rukletsov commented:
> The changes are fine, but is there an easy way for developers to look for sentinal values, especially if they are scattered in different pkgs?

For global sentinels — for sure, but you already know that : ). For package-level sentinels, we can say that those changing or using a package shall be able to discover errors defined in that package but I think your comment is valid. Maybe we can have a gentlemen's agreement (and a linter) that all package sentinels *must* be defined in `errors.go` file in that package?
